pySDC/implementations/problem_classes/AdvectionEquation_ND_FD.py

```diff
@@ -48,91 +48,105 @@ def __init__(
+            if (nvar + 1) % 2 != 0 and bc == 'dirichlet-zero':
                 raise ProblemError('setup requires nvars = 2^p - 1')
+        if ndim > 1 and nvars[1:] != nvars[:-1]:
+            raise ProblemError('need a square domain, got %s' % nvars)
 
         # invoke super init, passing number of dofs, dtype_u and dtype_f
-        super(advectionNd, self).__init__(
+        super().__init__(
             init=(nvars, None, np.dtype('float64')),
             dtype_u=mesh,
             dtype_f=mesh,
-            nvars=nvars,
-            c=c,
-            freq=freq,
-            stencil_type=stencil_type,
-            order=order,
-            lintol=lintol,
-            liniter=liniter,
-            direct_solver=direct_solver,
-            bc=bc,
-            ndim=ndim,
         )
 
-        if self.params.ndim == 1:
-            if type(self.params.nvars) is not tuple:
-                self.params.nvars = (self.params.nvars,)
-            if type(self.params.freq) is not tuple:
-                self.params.freq = (self.params.freq,)
-
         # compute dx (equal in both dimensions) and get discretization matrix A
-        if self.params.bc == 'periodic':
-            self.dx = 1.0 / self.params.nvars[0]
-            xvalues = np.array([i * self.dx for i in range(self.params.nvars[0])])
-        elif self.params.bc == 'dirichlet-zero':
-            self.dx = 1.0 / (self.params.nvars[0] + 1)
-            xvalues = np.array([(i + 1) * self.dx for i in range(self.params.nvars[0])])
+        if bc == 'periodic':
+            xvalues = np.linspace(0, 1, num=nvars[0], endpoint=False)
+        elif bc == 'dirichlet-zero':
+            xvalues = np.linspace(0, 1, num=nvars[0] + 2)[1:-1]
         else:
             raise ProblemError(f'Boundary conditions {self.params.bc} not implemented.')
+        dx = xvalues[1] - xvalues[0]
 
         self.A = problem_helper.get_finite_difference_matrix(
             derivative=1,
-            order=self.params.order,
-            stencil_type=self.params.stencil_type,
-            dx=self.dx,
-            size=self.params.nvars[0],
-            dim=self.params.ndim,
-            bc=self.params.bc,
+            order=order,
+            stencil_type=stencil_type,
+            dx=dx,
+            size=nvars[0],
+            dim=ndim,
+            bc=bc,
         )
-        self.A *= -self.params.c
+        self.A *= -c
+
+        self.xvalues = xvalues
+        self.Id = sp.eye(np.prod(nvars), format='csc')
+
+        # store relevant attributes
+        self.freq, self.sigma = freq, sigma
+        self.lintol, self.liniter, self.direct_solve = lintol, liniter, direct_solver
+
+        # read-only attributes
+        self._readOnly = {
+            name: eval(name)
+            for name in ['ndim', 'c', 'stencil_type', 'order']}
+
+        # register parameters
+        self._register('nvars', 'c', 'freq', 'stencil_type', 'order', 'lintol', 'liniter', 'direct_solver', 'bc')
```

Comment:
Wait, when we register `'c'` here and also have it as a readonly parameter, that is double, no?
I wonder if it would be any nicer to extend the register params class to be able to register readonly parameters.
Then you could do 
```
self._register('nvars', 'freq', 'lintol', 'liniter', 'direct_solver', 'bc')
self._register_readonly({'c': c, 'stencil_type': stencil_type, 'order': order})
```
and we could then overload the `__getattr__` method to streamline the readonly stuff a bit.

However, I wonder what's the difference between readonly and regular parameters. Neither the order, nor the boundary conditions should be allowed to be changed during the run, I feel.

Reply:
Register here just put parameters in params dictionnary, so they can be displayed later by the controller, or simply displayed all together. The readOnly parameter specifies argument for which any change will be ignored, for instance 'order' since changing it won't change the whole `A` matrix. Other arguments (like `sigma`, `lintol`, `lintiter`, etc ...) can be changed and there change would be taken into account in the object methods.

Overwriting the __getattr__ method is usually not recommended, since it add some overlay each time you want to access an attribute or use method. But I may have an idea that I can try ...

Reply:
Ok, found a way with the new commit. This can be just the standard approach when inheriting from a base class and adding supplementary parameters. First one has to store every parameters as a class attribute. Then register those parameters so they can be listed in the params property, and (optional) secure some parameters as read-only. What do you think ?